### PR TITLE
Improve auto lod by comparing every attribute

### PR DESCRIPTION
**Pull request by fire**
<WIP>

Fixes for some of the issues in https://github.com/godotengine/godot/issues/57416

- [ ] a struct to a float indexing change.
- [x] always use 8 bones / 8 weights.
- [x] Add a ui for changing lod indices

### CODE REVIEW

**Comment by fire** [comment]
@akien-mga do you think a pr that only increases the minimum to 256 triangle and or exposes the setting would be acceptable for 4.0?

**Comment by akien-mga** [comment]
I don't know, that depends on the scope of the actual change and whether it's tested to fix reported critical issues.

**Comment by clayjohn** [comment]
> @akien-mga do you think a pr that only increases the minimum to 256 triangle and or exposes the setting would be acceptable for 4.0?

IMO this is definitely 4.1 material. LOD changes are support difficult for us to test as the surface area for testing is huge.

**Comment by jcostello** [comment]
> > @akien-mga do you think a pr that only increases the minimum to 256 triangle and or exposes the setting would be acceptable for 4.0?
> 
> IMO this is definitely 4.1 material. LOD changes are support difficult for us to test as the surface area for testing is huge.

Can this break projects between 4.0 and 4.1?

**Comment by fire** [comment]
@clayjohn the only alternative that I can think of is set autolod to off and mark as experimental?

**Comment by realkotob** [comment]
This fixes critical issues which is why I think it should go into 4.0, otherwise practically every user importing animated models will have broken results.

The other alternative as mentioned is to have autolod be off by default and marked as experiment, but in that case then this should also be merged anyway.

**Comment by fire** [comment]
Out of band conversation. We agreed to have no change.

File: editor/import/resource_importer_scene.cpp [hunk 1]
```diff
@@ -52,6 +52,7 @@
 #include "scene/resources/sphere_shape_3d.h"
 #include "scene/resources/surface_tool.h"
 #include "scene/resources/world_boundary_shape_3d.h"
+#include <math.h>
```
Comment (fire) [inline editor/import/resource_importer_scene.cpp:55]: Remove.

Comment (fire) [inline editor/import/resource_importer_scene.cpp:55]:
```suggestion
```

**Comment by clayjohn** [comment]
I tested this on the TPS demo a few weeks ago and found that this resulted in almost no LODs being generated. Additionally the base LOD had way more vertices. The net effect was that the scene consistently had way more geometry than needed, even if LOD bias was set to always use the worst quality

**Comment by fire** [comment]
Can you provide some test case?

**Comment by clayjohn** [comment]
> Can you provide some test case?

All I did is re-import the main meshes in the 4.0 branch of the TPS-demo https://github.com/godotengine/tps-demo. You can reproduce my finding with any one of the meshes in the project (i.e. just reimport the robot if you want to save time)

**Comment by fire** [comment]
@clayjohn 

```
<param index="3" name="minimum_index_count" type="int" />
<param index="4" name="maximum_mesh_error" type="float" />
```

Thoughts on splitting this pr and adding hard overrides to the lod style with minimum_index_count and maximum_mesh_error.

**Comment by clayjohn** [comment]
> @clayjohn
> 
> ```
> <param index="3" name="minimum_index_count" type="int" />
> <param index="4" name="maximum_mesh_error" type="float" />
> ```
> 
> Thoughts on extracting this pr to add hard overrides to the lod import with minimum_index_count and maximum_mesh_error.

Sounds like it could be helpful. But we definitely need to investigate a bit more. I have a feeling we should be able to produce better meshes with the default settings.

**Comment by fire** [comment]
I tried a few of the test cases and after the AABB test fixes they're not failing anymore.

Can reopen if there's a need.

**Pull request closed without merging**